kv: TxnCoordSender: close of closed channel

From Gitter (user a6802739). Maybe we can spot the race easily.

```
> I0216 00:00:30.720429 58552 storage/replica_command.go:1300  initiating a split of range=1 [/Min-/Table/11) at key /System/"tsd!cr.store.valcount\x00\x01\x89\xf8\x06+`9"
I0216 00:00:35.797282 58552 server/status/runtime.go:137  runtime stats: 140 goroutines, 5.10MiB active, 524.20cgo/sec, 1.16/0.00 %(u/s)time, 0.00 %gc (3x)
I0216 00:00:45.537755 58552 storage/replica_command.go:1227  range 5: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1455552046.534s
panic: close of closed channel

goroutine 44 [running]:
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).cleanupTxn(0xc8203741e0, 0x7f1063e035f8, 0x23db298, 0xc82049bea0, 0xc820528540, 0x28, 0x30, 0x0, 0x143328dc2f3219fc, 0x0, ...)
    /root/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:523 +0x188
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).updateState(0xc8203741e0, 0x7f1063e03660, 0xc8206c79e0, 0x143328dc2f3219fc, 0x1, 0x0, 0x0, 0x0, 0x0, 0xc8209463b8, ...)
    /root/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:796 +0x2c5
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).Send(0xc8203741e0, 0x7f1063e03660, 0xc8206c79e0, 0x143328dc2f3219fc, 0x1, 0x0, 0x0, 0x0, 0x0, 0xc8209463b8, ...)
    /root/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:418 +0xe4a
github.com/cockroachdb/cockroach/client.(*txnSender).Send(0xc820946360, 0x7f1063e035b8, 0xc820013ac0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc8209463b8, ...)
    /root/go/src/github.com/cockroachdb/cockroach/client/txn.go:54 +0xca
github.com/cockroachdb/cockroach/client.(*DB).send(0xc820946360, 0xc820774da8, 0x1, 0x1, 0x91be9b, 0xc8207a77a0)
    /root/go/src/github.com/cockroachdb/cockroach/client/db.go:482 +0x16b
github.com/cockroachdb/cockroach/client.(*Txn).send(0xc820946360, 0xc820774da8, 0x1, 0x1, 0xc82098a000, 0xc8201570e0)
    /root/go/src/github.com/cockroachdb/cockroach/client/txn.go:501 +0x9d4
github.com/cockroachdb/cockroach/client.(*Txn).sendEndTxnReq(0xc820946360, 0xc820a70600, 0x0, 0x10)
    /root/go/src/github.com/cockroachdb/cockroach/client/txn.go:382 +0x1fa
github.com/cockroachdb/cockroach/client.(*Txn).Rollback(0xc820946360, 0xc8202d3ba0)
    /root/go/src/github.com/cockroachdb/cockroach/client/txn.go:378 +0x2f
github.com/cockroachdb/cockroach/client.(*Txn).Cleanup(0xc820946360, 0xc8206c7740)
    /root/go/src/github.com/cockroachdb/cockroach/client/txn.go:327 +0x3b
github.com/cockroachdb/cockroach/client.(*Txn).exec(0xc820946360, 0xc820775318, 0x15)
    /root/go/src/github.com/cockroachdb/cockroach/client/txn.go:439 +0x67b
github.com/cockroachdb/cockroach/client.(*DB).Txn(0xc82036c5a0, 0xc820775318, 0xc820845328)
    /root/go/src/github.com/cockroachdb/cockroach/client/db.go:463 +0x157
github.com/cockroachdb/cockroach/storage.(*Replica).AdminSplit(0xc8205561a0, 0x23db298, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /root/go/src/github.com/cockroachdb/cockroach/storage/replica_command.go:1302 +0xdee
github.com/cockroachdb/cockroach/storage.(*Replica).addAdminCmd(0xc8205561a0, 0x7f1063e03660, 0xc8206c7410, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /root/go/src/github.com/cockroachdb/cockroach/storage/replica.go:795 +0x565
github.com/cockroachdb/cockroach/storage.(*Replica).Send(0xc8205561a0, 0x7f1063e03660, 0xc8206c7410, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /root/go/src/github.com/cockroachdb/cockroach/storage/replica.go:648 +0x157
github.com/cockroachdb/cockroach/client.SendWrappedWith(0x7f1062578fd0, 0xc8205561a0, 0x7f1063e03660, 0xc8206c7410, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /root/go/src/github.com/cockroachdb/cockroach/client/sender.go:91 +0x1a6
```

--- a/kv/txn_coord_sender.go
+++ b/kv/txn_coord_sender.go
@@ -342,7 +342,7 @@ func (tc *TxnCoordSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*r
 			tc.Unlock()
 			if !ok {
 				pErr := roachpb.NewErrorf("writing transaction timed out, was aborted, " +
-					" or ran on multiple coordinators")
+					"or ran on multiple coordinators")
 				return nil, pErr
 			}
 		}
@@ -522,8 +522,9 @@ func (tc *TxnCoordSender) cleanupTxn(ctx context.Context, txn roachpb.Transactio
 	tc.Lock()
 	defer tc.Unlock()
 	txnMeta, ok := tc.txns[*txn.ID]
-	// The heartbeat might've already removed the record.
-	if !ok {
+	// The heartbeat might've already removed the record. Or we may have already
+	// closed txnEnd but we are racing with the heartbeat cleanup.
+	if !ok || txnMeta.txnEnd == nil {
 		return
 	}
 
@@ -551,7 +552,7 @@ func (tc *TxnCoordSender) unregisterTxnLocked(txnID uuid.UUID) (duration, restar
 
 	delete(tc.txns, txnID)
 
-	return
+	return duration, restarts, status
 }
 
 // heartbeatLoop periodically sends a HeartbeatTxn RPC to an extant

--- a/kv/txn_coord_sender_test.go
+++ b/kv/txn_coord_sender_test.go
@@ -641,6 +641,30 @@ func TestTxnCoordSenderTxnUpdatedOnError(t *testing.T) {
 	}
 }
 
+// TestTxnCoordIdempotentCleanup verifies that cleanupTxn is idempotent.
+func TestTxnCoordIdempotentCleanup(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	s := createTestDB(t)
+	defer s.Stop()
+	defer teardownHeartbeats(s.Sender)
+
+	txn := client.NewTxn(*s.DB)
+	ba := txn.NewBatch()
+	ba.Put(roachpb.Key("a"), []byte("value"))
+	if pErr := txn.Run(ba); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	s.Sender.cleanupTxn(context.Background(), txn.Proto)
+
+	ba = txn.NewBatch()
+	ba.InternalAddRequest(&roachpb.EndTransactionRequest{})
+	pErr := txn.Run(ba)
+	if pErr != nil && !testutils.IsPError(pErr, "aborted") {
+		t.Fatal(pErr)
+	}
+}
+
 // TestTxnMultipleCoord checks that a coordinator uses the Writing flag to
 // enforce that only one coordinator can be used for transactional writes.
 func TestTxnMultipleCoord(t *testing.T) {

